pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -19,12 +19,172 @@ package kuberuntime
+	command, args := kubecontainer.ExpandContainerCommandAndArgs(container, opts.Envs)
+	containerLogsPath := getContainerLogsPath(containerName, string(pod.UID))
+	podHasSELinuxLabel := pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.SELinuxOptions != nil
+	config := &runtimeApi.ContainerConfig{
+		Name:        &containerName,
+		Image:       &runtimeApi.ImageSpec{Image: &container.Image},
+		Command:     command,
+		Args:        args,
+		WorkingDir:  &container.WorkingDir,
+		Labels:      newContainerLabels(container, pod),
+		Annotations: newContainerAnnotations(container, pod, restartCount),
+		Mounts:      makeMounts(cid, opts, container, podHasSELinuxLabel),
+		LogPath:     &containerLogsPath,
+		Stdin:       &container.Stdin,
+		StdinOnce:   &container.StdinOnce,
+		Tty:         &container.TTY,
+	}
+
+	memoryLimit := container.Resources.Limits.Memory().Value()
+	cpuRequest := container.Resources.Requests.Cpu()
+	cpuLimit := container.Resources.Limits.Cpu()
+	var cpuShares int64
+	if cpuRequest.IsZero() && !cpuLimit.IsZero() {
+		cpuShares = milliCPUToShares(cpuLimit.MilliValue())
+	} else {
+		// if cpuRequest.Amount is nil, then milliCPUToShares will return the minimal number
+		// of CPU shares.
+		cpuShares = milliCPUToShares(cpuRequest.MilliValue())
+	}
+	if cpuShares != 0 || memoryLimit != 0 || m.cpuCFSQuota {
+		linuxResource := &runtimeApi.LinuxContainerResources{}
+		if cpuShares != 0 {
+			linuxResource.CpuShares = &cpuShares
+		}
+		if memoryLimit != 0 {
+			linuxResource.MemoryLimitInBytes = &memoryLimit
+		}
+		if m.cpuCFSQuota {
+			// if cpuLimit.Amount is nil, then the appropriate default value is returned
+			// to allow full usage of cpu resource.
+			cpuQuota, cpuPeriod := milliCPUToQuota(cpuLimit.MilliValue())
+			linuxResource.CpuQuota = &cpuQuota
+			linuxResource.CpuPeriod = &cpuPeriod
+		}
+
+		config.Linux = &runtimeApi.LinuxContainerConfig{
+			Resources: linuxResource,
+		}
+	}
+
+	if container.SecurityContext != nil {
+		securityContext := container.SecurityContext
+		if securityContext.Privileged != nil {
+			config.Privileged = securityContext.Privileged
+		}
+		if securityContext.ReadOnlyRootFilesystem != nil {
+			config.ReadonlyRootfs = securityContext.ReadOnlyRootFilesystem
+		}
+
+		if securityContext.Capabilities != nil {
+			if config.Linux == nil {
+				config.Linux = &runtimeApi.LinuxContainerConfig{
+					Capabilities: &runtimeApi.Capability{
+						AddCapabilities:  make([]string, 0, len(securityContext.Capabilities.Add)),
+						DropCapabilities: make([]string, 0, len(securityContext.Capabilities.Drop)),
+					},
+				}
+			}
+
+			for index, value := range securityContext.Capabilities.Add {
+				config.Linux.Capabilities.AddCapabilities[index] = string(value)
+			}
+			for index, value := range securityContext.Capabilities.Drop {
+				config.Linux.Capabilities.DropCapabilities[index] = string(value)
+			}
+		}
+
+		if securityContext.SELinuxOptions != nil {
+			if config.Linux == nil {
```

Comment:
Drop this if the above if is removedand so config.Linux is unconditionally initialized.